docs/proposals/20230407-managed-k8s-capi-contract-changes.md

```diff
@@ -181,9 +182,135 @@ As a cluster operator, I want to use Cluster API to provision and manage the lif
 
 As a service provider I want to be able to offer Managed Kubernetes clusters by using CAPI referencing my own managed control plane implementation that satisfies Cluster API contracts.
 
-### Current State of Managed Kubernetes in CAPI
+### Design
+
+Below we are documenting API changes and controllers changes required to omit the `<Infra>Cluster` entirely and to allow the `ControlPlane Provider` component  to take ownership of the responsibility of creating the control plane endpoint.
+
+#### Core Cluster API changes
+
+This proposal does not introduce any breaking changes for the existing "core" API. More specifically:
+
+The existing Cluster API types are already able to omit the `<Infra>Cluster`:
+
+- The `infrastructureRef` field on the Cluster object is already a pointer and thus it could be set to nil, and in fact we are already creating Clusters without `infrastructureRef` when we use a cluster class).
+- The `infrastructure.Ref` field on the ClusterClass objects already a pointer and thus it could be set to nil, but in this case it is required to change the validation webhook to allow the user to not specify it; on top of that, when validating inline patches, we should reject patches targeting the infrastructure template objects if not specified.
+
+In order to allow the `ControlPlane Provider` component  to take ownership of the responsibility of creating the control plane endpoint we are going to introduce a new `ClusterEndpoint` CRD, below some example:
```

Comment:
Would this `ClusterEndpoint` solve for https://github.com/kubernetes-sigs/cluster-api/issues/5295?

Reply:
@vincepri I think it's worth adding a reference to #5295 in the PR description. Do we feel like this proposal describes an implementation that, once implemented, will allow for others to build multiple controlplaneendpoint resources?

I think the answer is yes as this proposal specifies a generic endpoint spec, and describes them as a set. But WDYT?

Reply:
Stepping stone, but yes let's add a ref